---

<h1 style="text-align:center">THÍ NGHIỆM</h1>

---

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from Genetic_Algorithm import * 

### Nhập số liệu tính toán:

In [2]:
items, max_capacity = get_user_input()
get_info()


Thông tin bài toán:

Vật phẩm 1: Weight = 330, Value = 2870
Vật phẩm 2: Weight = 176, Value = 308
Vật phẩm 3: Weight = 2431, Value = 570
Vật phẩm 4: Weight = 2808, Value = 137
Vật phẩm 5: Weight = 1038, Value = 445
Vật phẩm 6: Weight = 3996, Value = 771
Vật phẩm 7: Weight = 3193, Value = 108
Vật phẩm 8: Weight = 805, Value = 469
Vật phẩm 9: Weight = 627, Value = 3678
Vật phẩm 10: Weight = 2257, Value = 2760
Vật phẩm 11: Weight = 1649, Value = 842
Vật phẩm 12: Weight = 1210, Value = 1237
Vật phẩm 13: Weight = 2642, Value = 1829
Vật phẩm 14: Weight = 508, Value = 967
Vật phẩm 15: Weight = 1192, Value = 313
Vật phẩm 16: Weight = 3446, Value = 1322
Vật phẩm 17: Weight = 3626, Value = 651
Vật phẩm 18: Weight = 1706, Value = 679
Vật phẩm 19: Weight = 3068, Value = 2254
Vật phẩm 20: Weight = 2977, Value = 525
Vật phẩm 21: Weight = 1688, Value = 1571
Vật phẩm 22: Weight = 709, Value = 697
Vật phẩm 23: Weight = 3522, Value = 2262
Vật phẩm 24: Weight = 85, Value = 3065
Vật phẩm 25: Weight = 335

## So sánh kĩ thuật:

### - Elitism Selection và Rank Selection

In [3]:
def select_population_rank(population):
    size_to_select = min(POPULATION_SIZE, len(population)) 
    ranked_population = sorted(population, key=lambda x: fitness(x), reverse=True) 
    total_rank = sum(range(1, len(population) + 1))
    rank_probabilities = [(len(population) - rank) / total_rank for rank in range(len(population))]
    selected_population = []
    for _ in range(size_to_select):  
        pick = random.uniform(0, 1)
        cumulative_probability = 0
        for i, prob in enumerate(rank_probabilities):
            cumulative_probability += prob
            if pick <= cumulative_probability:
                selected_population.append(ranked_population[i])
                break 
    return selected_population

# Danh sách các hàm chọn lọc
selection_strategies = {
    "Elitism Selection": select_population,  
    "Rank Selection": select_population_rank  
}

experiment_results = {}

def run_experiment_with_selection(selection_func):
    global fitness_history, select_population

    # Lưu giá trị hàm select_population ban đầu
    original_select_population = select_population

    # Thay thế hàm select_population cho thử nghiệm
    select_population = selection_func

    # Thực hiện thuật toán di truyền
    best_solution, best_fitness = genetic_algorithm()

    # Lưu kết quả
    result = {
        'best_solution': best_solution,
        'best_fitness': best_fitness,
        'fitness_history': fitness_history.copy()  # Lưu lại lịch sử fitness
    }

    # Reset fitness_history sau mỗi lần thử nghiệm
    fitness_history.clear()

    # Khôi phục hàm select_population ban đầu
    select_population = original_select_population

    return result

# Thực hiện các thử nghiệm cho từng chiến lược chọn lọc
for strategy_name, selection_func in selection_strategies.items():
    print(f"\nThử nghiệm với chiến lược chọn lọc: {strategy_name}")
    
    # Chạy thử nghiệm
    result = run_experiment_with_selection(selection_func)
    
    # Lưu kết quả vào experiment_results
    experiment_results[strategy_name] = result
    
    print(f"Best Solution: {result['best_solution']}")
    print(f"Best Fitness: {result['best_fitness']}")
    print(f"Fitness History: {result['fitness_history']}")



Thử nghiệm với chiến lược chọn lọc: Elitism Selection
Best Solution: [1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0]
Best Fitness: 52775
Fitness History: [37291, 38774, 38774, 38827, 42355, 42759, 42759, 44678, 45440, 45440, 45577, 49325, 49325, 49325, 49325, 49325, 49325, 50719, 50719, 50719, 50523, 50523, 50523, 50694, 50978, 51267, 51592, 51592, 51592, 52139, 52139, 51777, 51932, 52246, 52246, 52246, 52246, 52385, 52246, 52246, 52246, 52394, 52394, 52246, 52579, 52579, 52579, 52378, 52378, 52378, 52378, 52378, 52378, 52378, 52378, 52378, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775, 52775

### - Chạy nhiều lần để cho ra các thông số trung bình

In [27]:
import numpy as np

# Số lần chạy thử nghiệm
num_runs = 10

# Lưu kết quả tổng hợp
aggregated_results = {}

# Thực hiện thử nghiệm cho từng chiến lược chọn lọc
for strategy_name, selection_func in selection_strategies.items():
    fitness_final_list = []
    convergence_gen_list = []
    stability_std_list = []

    print(f"\nChiến lược chọn lọc: {strategy_name}")
    
    for run in range(num_runs):
        # Chạy thử nghiệm với chiến lược chọn lọc hiện tại
        result = run_experiment_with_selection(selection_func)
        history = result['fitness_history']
        final_fitness = history[-1]

        # Tính các giá trị thống kê
        fitness_final_list.append(final_fitness)
        convergence_gen = next((i for i, x in enumerate(history) if x >= 0.95 * final_fitness), len(history))
        convergence_gen_list.append(convergence_gen)
        stability_std = np.std(history)
        stability_std_list.append(stability_std)

    # Lưu kết quả trung bình cho chiến lược hiện tại
    aggregated_results[strategy_name] = {
        "average_fitness": np.mean(fitness_final_list),
        "average_convergence_gen": np.mean(convergence_gen_list),
        "average_stability_std": np.mean(stability_std_list),
    }

    # In ra kết quả
    print(f"- Fitness trung bình: {aggregated_results[strategy_name]['average_fitness']:.2f}")
    print(f"- Tốc độ hội tụ trung bình: {aggregated_results[strategy_name]['average_convergence_gen']:.2f}")
    print(f"- Độ ổn định trung bình: {aggregated_results[strategy_name]['average_stability_std']:.2f}")



Chiến lược chọn lọc: Elitism Selection
- Fitness trung bình: 52717.40
- Tốc độ hội tụ trung bình: 15.20
- Độ ổn định trung bình: 2052.56

Chiến lược chọn lọc: Rank Selection
- Fitness trung bình: 52635.40
- Tốc độ hội tụ trung bình: 15.40
- Độ ổn định trung bình: 2142.07


## So sánh kĩ thuật:

### - One-point crossover và Arithmetic Crossover

In [4]:
def crossover_arithmetic(parent1, parent2, alpha=0.5):
    child1 = []
    child2 = []
    for p1, p2 in zip(parent1, parent2):
        c1 = alpha * p1 + (1 - alpha) * p2
        c2 = (1 - alpha) * p1 + alpha * p2
        c1 = round(c1) if c1 in [0, 1] else random.randint(0, 1)
        c2 = round(c2) if c2 in [0, 1] else random.randint(0, 1)
        child1.append(c1)
        child2.append(c2)
    return child1, child2

# Danh sách các hàm lai ghép
crossover_strategies = {
    "One-point Crossover": crossover,  # Lai ghép một điểm
    "Arithmetic Crossover": crossover_arithmetic  # Lai ghép toán học
}

experiment_results = {}

# Hàm thực hiện thử nghiệm với từng chiến lược lai ghép
def run_experiment_with_crossover(crossover_func):
    global fitness_history, crossover

    # Lưu giá trị hàm crossover ban đầu
    original_crossover = crossover

    # Thay thế hàm crossover cho thử nghiệm
    crossover = crossover_func

    # Thực hiện thuật toán di truyền
    best_solution, best_fitness = genetic_algorithm()

    # Lưu kết quả
    result = {
        'best_solution': best_solution,
        'best_fitness': best_fitness,
        'fitness_history': fitness_history.copy()  # Lưu lại lịch sử fitness
    }

    # Reset fitness_history sau mỗi lần thử nghiệm
    fitness_history.clear()

    # Khôi phục hàm crossover ban đầu
    crossover = original_crossover

    return result

# Thực hiện các thử nghiệm cho từng chiến lược lai ghép
for strategy_name, crossover_func in crossover_strategies.items():
    print(f"\nThử nghiệm với chiến lược lai ghép: {strategy_name}")
    
    # Chạy thử nghiệm
    result = run_experiment_with_crossover(crossover_func)
    
    # Lưu kết quả vào experiment_results
    experiment_results[strategy_name] = result
    
    print(f"Best Solution: {result['best_solution']}")
    print(f"Best Fitness: {result['best_fitness']}")
    print(f"Fitness History: {result['fitness_history']}")



Thử nghiệm với chiến lược lai ghép: One-point Crossover
Best Solution: [1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1]
Best Fitness: 52557
Fitness History: [40939, 40720, 40765, 45539, 45539, 45539, 45539, 47715, 48869, 48869, 48869, 48869, 48869, 48869, 48869, 48869, 48869, 49947, 50153, 49788, 49788, 50869, 50869, 51566, 51566, 51566, 51566, 51566, 51566, 51566, 51566, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52128, 52517, 52517, 52517, 52517, 52517, 52517, 52517, 52517, 52517, 52517, 52517, 52517, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 52557, 525

### - Chạy nhiều lần để cho ra các thông số trung bình

In [36]:
import numpy as np

# Số lần chạy thử nghiệm
num_runs = 10

# Lưu kết quả tổng hợp
aggregated_results = {}

# Thực hiện thử nghiệm cho từng chiến lược lai ghép
for strategy_name, crossover_func in crossover_strategies.items():
    fitness_final_list = []
    convergence_gen_list = []
    stability_std_list = []

    print(f"\nChiến lược lai ghép: {strategy_name}")
    
    for run in range(num_runs):
        # Chạy thử nghiệm với chiến lược lai ghép hiện tại
        result = run_experiment_with_crossover(crossover_func)
        history = result['fitness_history']
        final_fitness = history[-1]

        # Tính các giá trị thống kê
        fitness_final_list.append(final_fitness)
        convergence_gen = next((i for i, x in enumerate(history) if x >= 0.95 * final_fitness), len(history))
        convergence_gen_list.append(convergence_gen)
        stability_std = np.std(history)
        stability_std_list.append(stability_std)

    # Lưu kết quả trung bình cho chiến lược hiện tại
    aggregated_results[strategy_name] = {
        "average_fitness": np.mean(fitness_final_list),
        "average_convergence_gen": np.mean(convergence_gen_list),
        "average_stability_std": np.mean(stability_std_list),
    }

    # In ra kết quả
    print(f"- Fitness trung bình: {aggregated_results[strategy_name]['average_fitness']:.2f}")
    print(f"- Tốc độ hội tụ trung bình: {aggregated_results[strategy_name]['average_convergence_gen']:.2f}")
    print(f"- Độ ổn định trung bình: {aggregated_results[strategy_name]['average_stability_std']:.2f}")


Chiến lược lai ghép: One-point Crossover
- Fitness trung bình: 52769.30
- Tốc độ hội tụ trung bình: 17.00
- Độ ổn định trung bình: 2157.16

Chiến lược lai ghép: Arithmetic Crossover
- Fitness trung bình: 52723.70
- Tốc độ hội tụ trung bình: 16.00
- Độ ổn định trung bình: 2188.38


## So sánh kĩ thuật:

### - Bit-flip Mutation và Swap Mutation

In [21]:
def mutate_swap(individual):
    if random.random() <= MUTATION_RATE:
        idx1 = random.randint(0, len(individual) - 1)
        idx2 = idx1
        while idx2 == idx1:  
            idx2 = random.randint(0, len(individual) - 1)
        individual[idx1], individual[idx2] = individual[idx2], individual[idx1]
    return individual
# Danh sách các hàm đột biến
mutation_strategies = {
    "Bit-flip Mutation": mutate,  # Đột biến Bit-flip
    "Swap Mutation": mutate_swap  # Đột biến Swap
}

experiment_results = {}

# Hàm thực hiện thử nghiệm với từng chiến lược đột biến
def run_experiment_with_mutation(mutation_func):
    global fitness_history, mutate

    # Lưu giá trị hàm mutate ban đầu
    original_mutate = mutate

    # Thay thế hàm mutate cho thử nghiệm
    mutate = mutation_func

    # Thực hiện thuật toán di truyền
    best_solution, best_fitness = genetic_algorithm()

    # Lưu kết quả
    result = {
        'best_solution': best_solution,
        'best_fitness': best_fitness,
        'fitness_history': fitness_history.copy()  # Lưu lại lịch sử fitness
    }

    # Reset fitness_history sau mỗi lần thử nghiệm
    fitness_history.clear()

    # Khôi phục hàm mutate ban đầu
    mutate = original_mutate

    return result

# Thực hiện các thử nghiệm cho từng chiến lược đột biến
for strategy_name, mutation_func in mutation_strategies.items():
    print(f"\nThử nghiệm với chiến lược đột biến: {strategy_name}")
    
    # Chạy thử nghiệm
    result = run_experiment_with_mutation(mutation_func)
    
    # Lưu kết quả vào experiment_results
    experiment_results[strategy_name] = result
    
    print(f"Best Solution: {result['best_solution']}")
    print(f"Best Fitness: {result['best_fitness']}")
    print(f"Fitness History: {result['fitness_history']}")



Thử nghiệm với chiến lược đột biến: Bit-flip Mutation
Best Solution: [1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1]
Best Fitness: 52840
Fitness History: [40396, 40704, 42597, 42597, 42798, 44474, 45438, 45438, 46738, 46539, 48836, 48895, 49598, 49598, 50114, 50114, 51367, 51367, 51367, 51367, 52296, 52296, 52296, 52296, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52639, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840, 52840

### - Chạy nhiều lần để cho ra các thông số trung bình

In [38]:
import numpy as np

# Số lần chạy thử nghiệm
num_runs = 10

# Lưu kết quả tổng hợp
aggregated_results = {}

# Thực hiện thử nghiệm cho từng chiến lược đột biến
for strategy_name, mutation_func in mutation_strategies.items():
    fitness_final_list = []
    convergence_gen_list = []
    stability_std_list = []

    print(f"\nChiến lược đột biến: {strategy_name}")
    
    for run in range(num_runs):
        # Chạy thử nghiệm với chiến lược đột biến hiện tại
        result = run_experiment_with_mutation(mutation_func)
        history = result['fitness_history']
        final_fitness = history[-1]

        # Tính các giá trị thống kê
        fitness_final_list.append(final_fitness)
        convergence_gen = next((i for i, x in enumerate(history) if x >= 0.95 * final_fitness), len(history))
        convergence_gen_list.append(convergence_gen)
        stability_std = np.std(history)
        stability_std_list.append(stability_std)

    # Lưu kết quả trung bình cho chiến lược hiện tại
    aggregated_results[strategy_name] = {
        "average_fitness": np.mean(fitness_final_list),
        "average_convergence_gen": np.mean(convergence_gen_list),
        "average_stability_std": np.mean(stability_std_list),
    }

    # In ra kết quả
    print(f"- Fitness trung bình: {aggregated_results[strategy_name]['average_fitness']:.2f}")
    print(f"- Tốc độ hội tụ trung bình: {aggregated_results[strategy_name]['average_convergence_gen']:.2f}")
    print(f"- Độ ổn định trung bình: {aggregated_results[strategy_name]['average_stability_std']:.2f}")


Chiến lược đột biến: Bit-flip Mutation
- Fitness trung bình: 52778.10
- Tốc độ hội tụ trung bình: 16.90
- Độ ổn định trung bình: 2212.43

Chiến lược đột biến: Swap Mutation
- Fitness trung bình: 52594.20
- Tốc độ hội tụ trung bình: 14.00
- Độ ổn định trung bình: 1935.63


---
# Tổng Hợp
---

### 1. Các tình huống

- Chạy nhiều lần mỗi test case, từ 10 đến 20.
- Kết quả sau mỗi lần chạy toàn bộ 6 test case ở file Testcase.txt.
- Tính trung bình cộng toàn bộ thông số bao gồm: fitness cuối cùng, tốc độ hội tụ, độ ổn định sau số lần chạy tuỳ ý.


**Testcase 1:**
| Kỹ thuật ảnh hưởng | Fitness cuối cùng | Tốc độ hội tụ (thế hệ) | Độ ổn định (std) |
|---------------------|-------------------|-----------------------|------------------|
| Elitism Selection   | 55.9              | 2                     | 0.53             |
| Rank Selection      | 55.6              | 3                     | 0.5              |
|---------------------|-------------------|-----------------------|------------------|
| One-point Crossover | 55.9              | 2                     | 0.5              |
| Arithmetic Crossover| 55.7              | 2                     | 0.4              |
|---------------------|-------------------|-----------------------|------------------|
| Bit-flip Mutation   | 56                | 3                     | 0.65             |
| Swap Mutation       | 55.6              | 3                     | 0.56             |

<br>

**Testcase 2:**
| Kỹ thuật ảnh hưởng | Fitness cuối cùng | Tốc độ hội tụ (thế hệ) | Độ ổn định (std) |
|---------------------|-------------------|-----------------------|------------------|
| Elitism Selection   | 68.8              | 5                     | 1.35             |
| Rank Selection      | 68.2              | 6                     | 1.3              |
|---------------------|-------------------|-----------------------|------------------|
| One-point Crossover | 68.2              | 5                     | 1.18             |
| Arithmetic Crossover| 68.5              | 6                     | 0.97             |
|---------------------|-------------------|-----------------------|------------------|
| Bit-flip Mutation   | 68.5              | 5                     | 1.09             |
| Swap Mutation       | 68.4              | 6                     | 1.39             |

<br>

**Testcase 3:**
| Kỹ thuật ảnh hưởng | Fitness cuối cùng | Tốc độ hội tụ (thế hệ) | Độ ổn định (std) |
|---------------------|-------------------|-----------------------|------------------|
| Elitism Selection   | 3096.4            | 4                     | 37.03            |
| Rank Selection      | 3096.1            | 5                     | 38.89            |
|---------------------|-------------------|-----------------------|------------------|
| One-point Crossover | 3118.2            | 4                     | 41.92            |
| Arithmetic Crossover| 3118.2            | 5                     | 44.26            |
|---------------------|-------------------|-----------------------|------------------|
| Bit-flip Mutation   | 3120.3            | 4                     | 39.25            |
| Swap Mutation       | 3114.6            | 5                     | 42.45            |

<br>

**Testcase 4:**
| Kỹ thuật ảnh hưởng | Fitness cuối cùng | Tốc độ hội tụ (thế hệ) | Độ ổn định (std) |
|---------------------|-------------------|-----------------------|------------------|
| Elitism Selection   | 7593.5            | 6                     | 112.80           |
| Rank Selection      | 7539              | 6                     | 108.87           |
|---------------------|-------------------|-----------------------|------------------|
| One-point Crossover | 7606              | 7                     | 113.87           |
| Arithmetic Crossover| 7571              | 5                     | 105.08           |
|---------------------|-------------------|-----------------------|------------------|
| Bit-flip Mutation   | 7592              | 7                     | 115.32           |
| Swap Mutation       | 7560.5            | 6                     | 109.25           |

<br>

**Testcase 5:**
| Kỹ thuật ảnh hưởng | Fitness cuối cùng | Tốc độ hội tụ (thế hệ) | Độ ổn định (std) |
|---------------------|-------------------|-----------------------|------------------|
| Elitism Selection   | 78876.2           | 15                    | 2764.81          |
| Rank Selection      | 78858.6           | 15                    | 2724.75          |
|---------------------|-------------------|-----------------------|------------------|
| One-point Crossover | 78869.4           | 15                    | 2824.42          |
| Arithmetic Crossover| 78810             | 15                    | 2832.44          |
|---------------------|-------------------|-----------------------|------------------|
| Bit-flip Mutation   | 78900.6           | 14                    | 2841.56          |
| Swap Mutation       | 78809.2           | 15                    | 2943.95          |

<br>

**Testcase 6:**
| Kỹ thuật ảnh hưởng | Fitness cuối cùng | Tốc độ hội tụ (thế hệ) | Độ ổn định (std) |
|---------------------|-------------------|-----------------------|------------------|
| Elitism Selection   | 52717.4           | 15                    | 2052.56          |
| Rank Selection      | 52635.4           | 15                    | 2142.07          |
|---------------------|-------------------|-----------------------|------------------|
| One-point Crossover | 52769.3           | 17                    | 2157.16          |
| Arithmetic Crossover| 52723.7           | 16                    | 2188.38          |
|---------------------|-------------------|-----------------------|------------------|
| Bit-flip Mutation   | 52778.1           | 17                    | 2212.43          |
| Swap Mutation       | 52594.2           | 14                    | 1935.63          |

# Tình huống tổng quan:
Sau khi đã thí nghiệm với các thông số và thuật toán, ta chọn ra được:

### Thông số: 
- 
- 
- CROSSOVER_RATE = 0.8  # Tỷ lệ lai ghép
- 
### Kỹ thuật sử dụng:
- Elitism Selection
- One-point Crossover
- Bit-flip Mutation
Ta sẽ dùng các thông số, kỹ thuật này để giải quyết vấn đề về tối ưu bài toán.